astropy/table/index.py

```diff
@@ -815,14 +806,85 @@ def __getitem__(self, item):
             # item should be a list or ndarray of values
             rows = []
             for key in item:
-                rows.extend(index.find((key,)))
+                p = index.find((key,))
+                if len(p) == 0:
+                    raise KeyError('No matches found for key {0}'.format(key))
+                else:
+                    rows.extend(p)
+        return rows
+
+    def __getitem__(self, item):
+        """
+        Retrieve Table rows by value slice.
+
+        Parameters
+        ----------
+        item : column element, list, ndarray, slice or tuple
+            Can be a value of the table primary index, a list/ndarray
+            of such values, or a value slice (both endpoints are included).
+            If a tuple is provided, the first element must be
+            an index to use instead of the primary key, and the
+            second element must be as above.
+        """
+        rows = self._get_rows(item)
 
         if len(rows) == 0:  # no matches found
             raise KeyError('No matches found for key {0}'.format(item))
         elif len(rows) == 1:  # single row
             return self.table[rows[0]]
         return self.table[rows]
 
+    def __setitem__(self, key, value):
+        """
+        Assign Table row's by value slice.
+
+            Parameters
+            ----------
+            key : column element, list, ndarray, slice or tuple
+                  Can be a value of the table primary index, a list/ndarray
+                  of such values, or a value slice (both endpoints are included).
+                  If a tuple is provided, the first element must be
+                  an index to use instead of the primary key, and the
+                  second element must be as above.
+
+            value : New values of the row elements.
+                    Can be a list of tuples/lists to update the row.
```

Comment:
Indentation is still not correct (you can take a look at the other docstrings to see examples).